### ISSUE TITLE
Allow working on extracted features

### ISSUE THREAD

**Issue by ebennequin** [sicara/easy-few-shot-learning#97]
# Why

Currently, all proposed datasets and solutions involve inferring through the backbone "on the fly" (during training or testing). When the backbone is frozen (as is often the case in FSL), it puts a huge necessary overhead. We want something that will allow to:

1. Extract all features for all images of my dataset
2. Perform experiments using these pre-computed features

# What

- A class `FeaturesDataset(FewShotDataset)`
  - `FeaturesDataset.__get_item__()` gives items as their embeddings
  - no preconception on the shape of the embedding (can be 1-dim or x-dim)
  - fields:
    - `embeddings`: tensor of shape `(dataset_length, **embedding_dimension)` 
    - `labels`: list of ints with same length as `embeddings`
    - `class_names`: list of strings with length `len(np.unique(labels))`
  - 4 methods to populate the dataset:
    - `from_dataframe(pd.DataFrame)` takes a df with 2 columns `"embedding"` and `"class_name"` with embedding an ArrayLike (ndarray or tensor),  populates the dataset's fields from this df
     - optional: if the df as an `image_path` column, it will populate an additional `image_paths` field 
    - `from_dataframe_parquet(PathLike)` takes a parquet file, reads the df inside and calls `from_dataframe()` on it
    - `from_dict(dict[str, ArrayLike])`: takes a dict where each key is a `class_name` and each value is a ndarray or tensor with shape `(n_images_for_this_class, **embedding_dimension)`, and populates the dataset's fields from this dict
    - `from_dict_pickle(PathLike)` opens a pickle, check it fits the expected content, and calls `from_dict()` on the content
  - `__init__()` takes 4 mutually exclusive arguments `source_dataframe`, `source_dict`, `source_dict_pickle`, `source_dataframe_parquet`. Depending on which is filled, it will call one of the population methods. If none is provided, it will initialize and empty dataset

- a method `predict_embeddings(dataloader: DataLoader, model: nn.Module, output_parquet: Optional[PathLike]) -> DataFrame`
  - infer on the whole dataloader, returns a df in the expected shape of `FeaturesDataset.from_dataframe()`'s input
  - if `output_parquet` is specified, saves the DF as parquet

- `backbone` is now an optional argument for `FewShotClassifier`
  - depending on the method (work on 1-dim embeddings or OK with multi-dimensional), if `backbone` is not provided, will initialize it as `nn.Identity` or `nn.Flatten`
  - remove internal logic of `FewShotClassifier` for computing backbone's feature dimension: make it a mandatory argument for FSL methods that need it
  - update notebooks and docstring accordingly

- a notebook `work_directly_on_embeddings.ipynb` that goes through the process of saving / using embeddings

- updated README.md

- extensive tests
